Add algorithm for calculating Wildes polarization efficiencies

rbauststfc opened the pull request:
### Description of work

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
Adds new algorithm `PolarizationEfficienciesWildes` that calculates the polarization efficiencies for a two-flipper instrument setup according to the Wildes 2006 paper. This implementation is based on the implementation that POLREF are currently using in their reduction code.

It is one of three algorithms that will be written under the linked issue.

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->

Refs #35682. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->
See linked issue for detailed description of what was required, and see the algorithm documentation (or the appropriate sections of the Wildes 2006 paper, as referenced in the linked issue) for the equations that have been implemented.

Note that the errors currently use the default Mantid errors. This is discussed on the main issue and, as stated there, we've agreed with our scientists that this is what should be done for the first version of this algorithm.

I will be adding a system test at a later stage (but before closing the linked issue), potentially for the whole workflow rather than specifically for this algorithm.

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->
- Build the `docs-html` target and check that the algorithm documentation renders OK and is understandable.

- The following script can be used to test the algorithm functionality:

```
ws00 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+12")
ws01 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+1")
ws10 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+2")
ws11 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+10")

nonMag = GroupWorkspaces([ws00, ws01, ws10, ws11])

wsM00 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+6")
wsM01 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+0.2")
wsM10 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+0.3")
wsM11 = CreateSampleWorkspace(XUnit="Wavelength", NumBanks=1, BankPixelWidth=1, Function="User Defined", UserDefinedFunction="name=UserFunction, Formula=x*0+1")

mag = GroupWorkspaces([wsM00, wsM01, wsM10, wsM11])

PolarizationEfficienciesWildes('nonMag', 'mag', Flippers='00,01,10,11', IncludeDiagnosticOutputs=True, OutputFpEfficiency="fp", OutputFaEfficiency="fa", OutputPolarizerEfficiency="p", OutputAnalyserEfficiency="a")
```
- If you would like to test with real calibration runs then talk to Rachel.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

rbauststfc committed:
Add new algorithm boilerplate

Includes inputs, outputs and validation, with unit tests

rbauststfc committed:
Add main algorithm contents and unit tests

rbauststfc committed:
Refactor to use member variables instead of passing workspaces around

rbauststfc committed:
Add progress reporting

rbauststfc committed:
Add validation that the bin boundaries match across all input workspaces

rbauststfc committed:
Add and amend algorithm documentation and amend algorithm property descriptions

rbauststfc committed:
Add release notes for issue 35682

rbauststfc committed:
Improve validation code efficiency and robustness

sf1919 commented:
@rbauststfc please ignore the 200+ OSX test failures - problem with a machine. PR should run on a different machine next time you push a commit.

cailafinn reviewed (changes requested):
Code looks really good, just a few very minor comments. Docs looking good, just a note on the See Also section. 

Functionally tested using the provided script, which worked really well. Some further functional testing revealed an issue when either the `OutputPolarizerEfficiency` or `OutputAnalyzerEfficiency` workspace names are not provided. This only occurs in the algorithm dialog if it has been set to Keep open, run once with the flipper/polariser names set, and then run again with the names removed. 
```
Error in execution of algorithm PolarizationEfficienciesWildes:
Add Data Object with empty name
```

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizationEfficienciesWildes.cpp:
```diff
@@ -0,0 +1,407 @@
+  setProperty(PropNames::OUTPUT_F_P_EFF_WS, m_wsFp);
+  setProperty(PropNames::OUTPUT_F_A_EFF_WS, m_wsFa);
+
+  if (m_wsP != nullptr) {
```
Thanks, this is a good spot and a consequence I hadn't thought about when using member variables.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizationEfficienciesWildes.cpp:
```diff
@@ -0,0 +1,407 @@
+  // We use the flipper efficiency to multiply the mag ws counts, but the resulting workspace will have lost the Y unit
+  // and distribution information. We need to put these back otherwise the rest of the calculation fails when it tries
+  // to add and subtract workspaces with different Y units.
+  const auto twoFp = 2 * m_wsFp;
+  const auto twoFa = 2 * m_wsFa;
+
+  const auto twoFa00 = (1 - twoFa) * ws00;
+  setUnitAndDistributionToMatch(twoFa00, ws00);
+
+  const auto twoFa10 = (twoFa - 1) * ws10;
+  setUnitAndDistributionToMatch(twoFa10, ws10);
+
+  const auto twoFp00 = (1 - twoFp) * ws00;
+  setUnitAndDistributionToMatch(twoFp00, ws00);
+
+  const auto twoFp01 = (twoFp - 1) * ws01;
+  setUnitAndDistributionToMatch(twoFp01, ws01);
+
+  const auto numerator = twoFa00 + twoFa10 - ws01 + ws11;
+  const auto denominator = twoFp00 + twoFp01 - ws10 + ws11;
+  const auto tpmoSquared = m_wsPhi * (numerator / denominator);
```
I thought about the same thing following the useful insight you got on performance from the work you did on that other PR. On balance I decided against it here, though, because looking more closely at the add/multiply etc. Mantid methods I felt that they do some useful validation and include some other logic that I wasn't keen to either duplicate or go without in this particular case. On real data the algorithm takes approximately 30 seconds to run, which is acceptable for the purpose of calibration (it should be run a maximum of once per experiment, although as I understand it probably more like once per cycle, and likely less than that as currently POLREF are intending to continue using their own implementation, which runs at the same sort of speed). If performance ever becomes more of a consideration then we can come back to it at that stage.

It might be worth flagging your findings on performance of Mantid workspace arithmetic to the Core team/more generally, though. I think for long-term maintainability it would be better for a more central solution to be found to improve workspace arithmetic performance and then have that made available to all algorithms.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/PolarizationEfficienciesWildes.cpp:
```diff
@@ -0,0 +1,407 @@
+  auto makeSettingIncludeDiagnosticsIsSelected = [] {
+    return std::make_unique<Kernel::EnabledWhenProperty>(PropNames::INCLUDE_DIAGNOSTICS, IS_EQUAL_TO, "1");
+  };
+  std::shared_ptr<IPropertySettings> setting =
+      std::make_shared<Kernel::EnabledWhenProperty>(PropNames::INCLUDE_DIAGNOSTICS, IS_EQUAL_TO, "1");
```
Ah, good spot, something from a previous iteration that I must have forgotten to remove.

rbauststfc commented on docs/source/algorithms/PolarizationCorrectionWildes-v1.rst:
```diff
@@ -153,6 +153,8 @@ References
 .. [#WILDES] A. R. Wildes, *Neutron News*, **17** 17 (2006)
              `doi: 10.1080/10448630600668738 <https://doi.org/10.1080/10448630600668738>`_
 
+.. seealso :: Algorithm :ref:`algm-PolarizationEfficienciesWildes`
+
```
I'd be lying if I said I really understand how this is working 🙂 I noticed something similar seems to be in use in other documentation i.e. https://github.com/mantidproject/mantid/blob/main/docs/source/algorithms/ReflectometryISISLoadAndProcess-v1.rst?plain=1#L122, but that one is a Python algorithm. I'll take another look and see if I can figure out how this needs to be set up.

cailafinn replied:
It's collected by the `.. relatedalgorithms::` directive at the top of the file. But including a `seealso::` overrides that and stops it from appearing properly. [The code for the directive](https://github.com/mantidproject/mantid/blob/main/docs/sphinxext/mantiddoc/directives/relatedalgorithms.py) basically just calls the agorithm's `seeAlso()` method and then sorts out the links.

rbauststfc replied:
Great, thanks, and do you know why it seems to be a bit different for Python algorithms? I noticed a few docs files for Python algorithms that seem to include the seealso line in the rst file?

cailafinn replied:
I think it's just a copying thing. For instance [HFIRSANS2Wavelength](https://docs.mantidproject.org/nightly/algorithms/HFIRSANS2Wavelength-v1.html) is a python algorithm that uses the `SeeAlso()` method and directive and works just like the C++ ones.

rbauststfc replied:
I found that the issue with the documentation in this PR was that neither had the `.. relatedalgorithms::` directive at the top, which as you say is what is actually required to pull through the `seeAlso()` algorithms. It didn't seem to make any difference whether I also included the `.. seealso` directive, it worked with or without (which explains why some `.rst` files that contain both still work, like the one we have for `ReflectometrISISLoadAndProcess`). I'll remove the `.. seealso` directive from the docs in this PR, though, because it's obsolete even if it may not be causing any issues.

cailafinn replied:
Thanks, I'd missed that I'd also added the line in the Wildes workspace order PR, that's good to know.

rbauststfc committed:
Don't return previous values when alg run twice

The algorithm uses private member variables that need to be cleared to
prevent previously calculated values being returned if an instance of
the algorithm is run twice. In addition, the algorithm provides some
optional diagnostic output properties that we provide default values for. If the algorithm is being run as a child
then these can incorrectly remain populated with previous values if we do not clear them when they are not required. It seems that setting the property value to it's current value (to deliberately retain any edits the user has made to that) is sufficient to clear the associated output workspace.

cailafinn reviewed (approved):
Changes to the code look good, functionally working well for multiple runs, documentation looking good. Nice work, happy to approve :shipit:
